File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListModule.java
package com.spacex.ui.rocketList;

import android.os.Handler;
import com.spacex.rockets.repository.RocketsRepository;
import com.spacex.rockets.repository.RocketsRepositoryModule;
import com.spacex.ui.ExecutorsModule;
import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import java.util.concurrent.Executor;

@Module(includes = {RocketsRepositoryModule.class, ExecutorsModule.class})
public abstract class RocketListModule {

  @Provides
  static RocketListContract.Presenter providePresenter(
      RocketListContract.View view, RocketListContract.Coordinator coordinator) {
    return new RocketListPresenter(view, coordinator);
  }

  @Provides
  static RocketListContract.Coordinator provideCoordinator(
      RocketsRepository rocketsRepository, Handler mainThreadHandler, Executor backgroundExecutor) {
    return new RocketListCoordinator(rocketsRepository, mainThreadHandler, backgroundExecutor);
  }

  @Binds
  abstract RocketListContract.View bindView(RocketListActivity rocketListActivity);
}

File: app/src/main/kotlin/com/spacex/ui/rocketList/RocketListModule.kt
package com.spacex.ui.rocketList

import android.os.Handler
import com.spacex.rockets.repository.RocketsRepository
import com.spacex.rockets.repository.RocketsRepositoryModule
import com.spacex.ui.ExecutorsModule
import dagger.Binds
import dagger.Module
import dagger.Provides
import java.util.concurrent.Executor

@Module(includes = [RocketsRepositoryModule::class, ExecutorsModule::class])
abstract class RocketListModule {

    @Binds
    abstract fun bindView(rocketListActivity: RocketListActivity): RocketListContract.View

    @Module
    companion object {

        @JvmStatic
        @Provides
        fun providePresenter(
                view: RocketListContract.View,
                coordinator: RocketListContract.Coordinator
        ): RocketListContract.Presenter = RocketListPresenter(view, coordinator)

        @JvmStatic
        @Provides
        fun provideCoordinator(
                rocketsRepository: RocketsRepository,
                mainThreadHandler: Handler,
                backgroundExecutor: Executor): RocketListContract.Coordinator =
                RocketListCoordinator(rocketsRepository, mainThreadHandler, backgroundExecutor)
    }
}
